plugins/Model_GAN128/Trainer.py

```diff
@@ -0,0 +1,222 @@
+        loss_DA, loss_GA = self.define_loss(self.model.netDA, real_A, fake_A, fake_sz64_A, distorted_A, vggface_feat)
+        loss_DB, loss_GB = self.define_loss(self.model.netDB, real_B, fake_B, fake_sz64_B, distorted_B, vggface_feat)
+
+        loss_GA += 3e-3 * K.mean(K.abs(mask_A))
+        loss_GB += 3e-3 * K.mean(K.abs(mask_B))
+
+        weightsDA = self.model.netDA.trainable_weights
+        weightsGA = self.model.netGA.trainable_weights
+        weightsDB = self.model.netDB.trainable_weights
+        weightsGB = self.model.netGB.trainable_weights
+
+        # Adam(..).get_updates(...)
+        training_updates = Adam(lr=self.lrD, beta_1=0.5).get_updates(weightsDA,[],loss_DA)
+        self.netDA_train = K.function([distorted_A, real_A],[loss_DA], training_updates)
+        training_updates = Adam(lr=self.lrG, beta_1=0.5).get_updates(weightsGA,[], loss_GA)
+        self.netGA_train = K.function([distorted_A, real_A], [loss_GA], training_updates)
+
+        training_updates = Adam(lr=self.lrD, beta_1=0.5).get_updates(weightsDB,[],loss_DB)
+        self.netDB_train = K.function([distorted_B, real_B],[loss_DB], training_updates)
+        training_updates = Adam(lr=self.lrG, beta_1=0.5).get_updates(weightsGB,[], loss_GB)
+        self.netGB_train = K.function([distorted_B, real_B], [loss_GB], training_updates)
+
+    def train_one_step(self, iter, viewer):
+        # ---------------------
+        #  Train Discriminators
+        # ---------------------
+
+        # Select a random half batch of images
+        epoch, warped_A, target_A = next(self.train_batchA)
+        epoch, warped_B, target_B = next(self.train_batchB)
+
+        # Train dicriminators for one batch
+        if iter % 1 == 0:
+            errDA  = self.netDA_train([warped_A, target_A])
+            errDB  = self.netDB_train([warped_B, target_B])
+
+        # Train generators for one batch
+        errGA = self.netGA_train([warped_A, target_A])
+        errGB = self.netGB_train([warped_B, target_B])
+        
+        print('[%s] [%d/%s][%d] Loss_DA: %f Loss_DB: %f Loss_GA: %f Loss_GB: %f'
+              % (time.strftime("%H:%M:%S"), epoch, "num_epochs", iter, errDA[0], errDB[0], errGA[0], errGB[0]),
+              end='\r')
+        
+        if viewer is not None:
+            self.show_sample(viewer)
+    
+    def cycle_variables(self, netG):
+        distorted_input = netG.inputs[0]
+        fake_output = netG.outputs[0]
+        fake_output64 = netG.outputs[1]
+        alpha = Lambda(lambda x: x[:,:,:, :1])(fake_output)
+        rgb = Lambda(lambda x: x[:,:,:, 1:])(fake_output)
+        
+        masked_fake_output = alpha * rgb + (1-alpha) * distorted_input 
+
+        fn_generate = K.function([distorted_input], [masked_fake_output])
+        fn_mask = K.function([distorted_input], [concatenate([alpha, alpha, alpha])])
+        fn_abgr = K.function([distorted_input], [concatenate([alpha, rgb])])

```

Comment:
@shaoanlu what would we use `fn_abgr` for?

Reply:
I used it for debugging and forgot to delete it (not sure if it is called somewhere in the script).